src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -213,32 +240,87 @@ dfs vs = concatMap flatten . dfsForestFrom vs
 -- 'isSubgraphOf' ('vertices' $ reachable x y) y == True
 -- @
 reachable :: Ord a => a -> AdjacencyMap a -> [a]
-reachable x = concat . bfs [x]
+reachable x = dfs [x]
+
+type Cycle = NonEmpty 
+data Entry a = Entered a | Exited a
+type TopSort a = Either (Cycle a) [a]
+type ParentTable a = Map.Map a (Entry a)
+data S a = S { table :: !(ParentTable a), order :: [a] }
+
+topSort' :: (Ord a, MonadState (S a) m, MonadCont m)
+         => AdjacencyMap a -> m (TopSort a)
+topSort' g = callCC $ \cyclic ->
+  do let vertices = map fst $ Map.toDescList $ adjacencyMap g
+         adjacent = Set.toDescList . flip postSet g
+         dfs prev curr =
+           do enter prev curr
+              forM_ (adjacent curr) $ \next ->
+                nodeState next >>= \case
+                  Nothing -> dfs curr next
+                  Just (Exited _) -> return ()
+                  _ -> cyclic . Left . retrace curr next =<< gets table
```

Comment:
If you expand the remaining pattern `Just (Entered _)` it would be a bit easier to read. I hope it doesn't come with any performance cost.

Reply:
In fact, if you had `Just (Entered v)` and traced the cycle starting from `v`, could it produce more intuitive results?

Reply:
In cases like `edge 4 4`, or `circuit [1..3]`, `Entered v` will indicate the root of the search tree, which was previously `undefined`. `next` should be `v` in general, I think. 

The current documentation for `topSort` may be too convoluted, but it should accurately describe which cycle is found first. I definitely agree that it's not intuitive, and could modify it to present the cycle from the smallest node or something like that...

Reply:
Now that you added new constructors, it can no longer be `undefined`, right? If yes, please expand all `_` patterns, as otherwise, it's always necessary to make the additional effort of remembering the remaining constructors when reading the code. (Also, if a new constructor is added in future such catch-all pattern matches will be difficult to notice.)

Reply:
Yes, the search now starts by `dfs OpenRoot v` when `v` is not yet in the search tree. 

There is still `_` for the case of a cycle, where otherwise 

```
                  Just (Entered v) -> cyclic . Left . retrace curr next =<< gets table
                  Just OpenRoot  -> cyclic . Left . retrace curr next =<< gets table
```

would need to be written. Shall I fill all such cases in?

Reply:
I see. If there is a repetition `cyclic . Left . retrace curr next =<< gets table` you can just give it a name, something like `findCycle` and use in both cases.

What about `Just ClosedRoot`? Is it unreachable? Then better raise an `error` to at least make it obvious for the reader that this case shouldn't happen.

Reply:
`ClosedRoot` is matched against in `dfs` in the current implementation, where it's possible. It's not possible when recovering the cycle though, but where we still have `_` to match that, and an error is raised. I'll just add complete pattern matches everywhere?

Reply:
Yes, I think it's just safer and clearer to use complete pattern matches everywhere.

Reply:
One small side effect of writing out full pattern matches is `-Wunused-matches` warnings during builds. Should that be suppressed by adding to the flags in the cabal file or something along those lines?

Reply:
> One small side effect of writing out full pattern matches is `-Wunused-matches` warnings during builds.

Could you clarify which pattern is unused? I've had a look at the code but all patterns seem necessary.

> Should that be suppressed by adding to the flags in the cabal file or something along those lines?

No, this is a useful warning. If a pattern is indeed unused we should delete it.

Reply:
Changing ones like `Just (Exited _)` to `Just (Exited v)` gives the warning, the `v` is unused. 

I guess what's confusing is that the parent table and the node state table are blended into one. We could instead have two maps (one for parents, and one for node state `Entered` &`Exited`) , but I believe performance will be slower

Reply:
There will be two map operations instead of one each time a vertex is visited, but I'm trying it out in case the performance hit isn't as much as I expect

Reply:
Well, my intuition was wrong! It's faster than before (on cases with cycles and without). I guess less indirection tends to help! I also find it more readable.

Reply:
Awesome, much nicer!